keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1067 @@
+* VolumeGroup controller will be triggered to update the VolumeGroup Status.
+* Deleting a PVC from a VolumeGroup will trigger external-provisioner and the VolumeGroup controller as well.
+
+#### Create VolumeGroup from VolumeGroupSnapshot
+
+Creating a new volume group from an existing group snapshot is supported if the CSI driver supports VOLUMEGROUP capability. As a result, PVCs will be created from source snapshots and placed in a new volume group.
+
+#### Create VolumeGroup with Selector
+
+Creating a new volume group and adding existing PVCs matching the label selector to the group is supported if the CSI driver supports VOLUMEGROUP capability.
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+### Create VolumeGroupSnapshot
+
+A VolumeGroupSnapshot can be created with a VolumeGroup as the source if the CSI driver supports the GROUPSNAPSHOT capability.
+* Create a VolumeGroupSnapshot with a VolumeGroup as the source.
+* This will trigger the VolumeGroupSnapshot controller to call the CreateVolumeGroupSnapshot CSI function and also create multiple VolumeSnapshot API objects with VolumeGroupSnapshot name parameter in each VolumeSnapshot Spec. This will trigger the creation of VolumeSnapshotContent API objects in the snapshot controller and calls to the CreateSnapshot CSI function in the CSI snapshotter sidecar. The CSI snapshotter sidecar will pass the new group_snapshot_name parameter to the CSI Driver when calling CreatSnapshot.
+* When CSI driver receives CreateSnapshot request for individual snapshots with a VolumeGroupSnapshot name:
+  * Case 1: If it knows how to create a group snapshot on the storage system, it returns (nil, nil), and leave it to the CreateVolumeGroupSnapshot function to handle the snapshot creation.
+  * Case 2: If it does not know how to create a group snapshot on the storage system, it will create an individual snapshot as usual and return the snapshot_id back.
+* CreateVolumeGroupSnapshot CSI function response
+  * Case 1: The CreateVolumeGroupSnapshot CSI function should return a list of snapshots (Snapshot message defined in CSI Spec) in its response. The VolumeGroupSnapshot controller can use the returned list of snapshots to update corresponding individual VolumeSnapshotContents, wait for VolumeSnapshots and VolumeSnapshotContents to be bound, and update SnapshotList in the VolumeGroupSnapshot Status.
+  * Case 2: The CreateVolumeGroupSnapshot CSI function returns group_snapshot_id and volume_group_id, but leaves snapshots field as empty. The VolumeGroupSnapshot controller watches VolumeSnapshot and VolumeSnapshotContent API objects. When VolumeSnapshot and VolumeSnapshotContent are bound, it saves the VolumeSnapshot API object to SnapshotList in its Status.
+
```

Comment:
Admissions control or finalizer to prevent individual snapshot from being deleted that belong to a GroupSnapshot.

Reply:
TODO: write up details on how this would work when there are multiple VolumeSnapshots/Contents along with a VolumeGroupSnapshot and how snapshot is mapped to PVC at restore time.  Can user provide PVC names or can they be retrieved from VolumeGroupSnapshot status and reconstructed at restore time?

Reply:
> TODO: write up details on how this would work when there are multiple VolumeSnapshots/Contents along with a VolumeGroupSnapshot and how snapshot is mapped to PVC at restore time. Can user provide PVC names or can they be retrieved from VolumeGroupSnapshot status and reconstructed at restore time?

VolumeGroupSnapshot status should have PVC to VolumeSnapshot mapping so that this can be reconstructed at restore time.